### PR TITLE
[CARBONDATA-2879] [CARBONDATA-2918] support sort scope for sdk and make sort column case insensitive

### PR DESCRIPTION
**Pull request by ajantha-bhat**
**problem:**: [CARBONDATA-2879] SDK doesn't support batch sort and no sort as we don't have any API support for sort scope

**Solution**: provide sort_scope in existing load_options, so user can decide which sort_scope. 
**currently supported batch_sort, local_sort and no_sort.**
global sort is not applicable in sdk scenario. Hence not supported.
Updated the method header and doc also

**problem:** [CARBONDATA-2918]
Currently in CarbonWriterBuilder,  
sortColumnsList.indexOf(field.getFieldName()) return null due to case issue. so, one of the sortcolumns in array will be null. so one of the column schema becomes null,
hence when column schema is accessed, we get NPE

**solution:** Change pareant column (as sort column is supported only foro parent columns) name to lower case when schema received by the user 
and change sort columns also to lower case when received by the user.

Be sure to do all of the following checklists to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed? NA
 
 - [ ] Any backward compatibility impacted? NA
 
 - [ ] Document update required? yes, updated.

 - [ ] Testing done.
Added UT
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. NA

### CODE REVIEW

**Comment by xuchuanyin** [comment]
I think you can go through all the load options and explicitly support or not support them.
Moreover, these options should also be considered for dataframe write.
Better to keep them consistent.

**Comment by ajantha-bhat** [comment]
@xuchuanyin : Already gone through all the load options and decided to support 9 options. This sort scope was a create table property. This we were not supporting. Now we are supporting. 
For SDK create table property and load options can be combined as all write() is considered for 1 load (1 writer, 1 load).

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8300/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/230/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTable.scala [hunk 1]
```diff
@@ -2359,6 +2359,46 @@ class TestNonTransactionalCarbonTable extends QueryTest with BeforeAndAfterAll {
     checkAnswer(sql("select * from sdkOutputTable"), Seq(Row(Timestamp.valueOf("1970-01-02 16:00:00"), Row(Timestamp.valueOf("1970-01-02 16:00:00")))))
   }
 
+  test("test Sort Scope with SDK") {
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTable.scala]: yes, later will be handled in separate PR.

we can move AVRO test cases to another file and keep only non-avro in this

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonWriterBuilder.java [hunk 2]
```diff
@@ -265,12 +262,20 @@ public CarbonWriterBuilder withLoadOptions(Map<String, String> options) {
...
+          !option.equalsIgnoreCase("escapechar") &&
+          !option.equalsIgnoreCase("sort_scope")) {
+        throw new IllegalArgumentException("Unsupported option:" + option
+            + ". Refer method header or documentation");
       }
     }
 
+    // validate sort scope
+    String sortScope = options.get("sort_scope");
+    if ((sortScope != null) && (!(sortScope.equalsIgnoreCase("local_sort")) && !(sortScope
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonWriterBuilder.java]: OK. I use this now.

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java [hunk 3]
```diff
@@ -59,7 +59,7 @@
     optionsFinal.put("fileheader", Maps.getOrDefault(options, "fileheader", ""));
     optionsFinal.put("commentchar", Maps.getOrDefault(options, "commentchar", "#"));
     optionsFinal.put("columndict", Maps.getOrDefault(options, "columndict", null));
-
+    optionsFinal.put("sort_scope", Maps.getOrDefault(options, "sort_scope", "local_sort"));
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8304/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/234/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8308/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/238/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/249/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8319/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8325/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/255/

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by ravipesala** [comment]
@jackylk I feel it is better to add sortscope to carbondata file. In future if we want to do compaction it will be easy to do merge sort for batch sorted and local sorted files. For no sort we need to sort and do merge sort or just appending it.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8327/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/257/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8329/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/259/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2/

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java [hunk 4]
```diff
@@ -59,7 +59,8 @@
     optionsFinal.put("fileheader", Maps.getOrDefault(options, "fileheader", ""));
     optionsFinal.put("commentchar", Maps.getOrDefault(options, "commentchar", "#"));
     optionsFinal.put("columndict", Maps.getOrDefault(options, "columndict", null));
-
+    optionsFinal.put("sort_scope",
+        Maps.getOrDefault(options, "sort_scope", CarbonCommonConstants.LOAD_SORT_SCOPE_DEFAULT));
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java]: Please check in the CrabonSession Load scenario we are restricting it or not. It is not supposed to use per load . It should be only for SDK

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java [hunk 5]
```diff
@@ -183,7 +184,6 @@
     }
 
     optionsFinal.put("single_pass", String.valueOf(singlePass));
-    optionsFinal.put("sort_scope", "local_sort");
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java]: Check for SDK flow

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonWriterBuilder.java [hunk 6]
```diff
@@ -311,15 +326,15 @@ public CarbonWriterBuilder withTableProperties(Map<String, String> options) {
...
+      if (entry.getKey().equalsIgnoreCase("blocksize")) {
         this.withBlockSize(Integer.parseInt(entry.getValue()));
       } else if (entry.getKey().equalsIgnoreCase("blockletsize")) {
         this.withBlockletSize(Integer.parseInt(entry.getValue()));
       } else if (entry.getKey().equalsIgnoreCase("localDictionaryThreshold")) {
         this.localDictionaryThreshold(Integer.parseInt(entry.getValue()));
       } else if (entry.getKey().equalsIgnoreCase("enableLocalDictionary")) {
         this.enableLocalDictionary((entry.getValue().equalsIgnoreCase("true")));
-      } else {
+      } else if (entry.getKey().equalsIgnoreCase("sortcolumns")) {
```
Comment (ravipesala) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonWriterBuilder.java]: Please make sure all property names are the same as create table properties

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java (hunk 4 above)
Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java]: yes, it should not be set here.

now moved to 
CarbonLoadModelBuilder.build() , so it will set only for SDK and spark-file-format

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java (hunk 5 above)
Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java]: same fix as above. reverted here.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonWriterBuilder.java (hunk 6 above)
Comment (ajantha-bhat) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonWriterBuilder.java]: yes. changed it to same as table properties syntax

**Comment by ajantha-bhat** [comment]
@ravipesala : All comments are handled. please check.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/127/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/295/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8365/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/128/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/296/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8366/

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/CarbonLoadModel.java [hunk 7]
```diff
@@ -62,6 +62,7 @@
   /* Number of thread in which sdk writer is used */
   private short sdkUserCores;
 
+
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/loading/model/CarbonLoadModel.java]: remove unnecessary changes from file

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java [hunk 8]
```diff
@@ -59,7 +59,6 @@
     optionsFinal.put("fileheader", Maps.getOrDefault(options, "fileheader", ""));
     optionsFinal.put("commentchar", Maps.getOrDefault(options, "commentchar", "#"));
     optionsFinal.put("columndict", Maps.getOrDefault(options, "columndict", null));
-
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java]: remove unnecessary changes from file

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/CarbonLoadModel.java (hunk 7 above)
Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/loading/model/CarbonLoadModel.java]: done. removed

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java (hunk 8 above)
Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java]: done. removed

**Comment by ajantha-bhat** [comment]
@ravipesala : OK. removed empty lines.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/136/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8374/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/304/

**Comment by ravipesala** [comment]
LGTM
